pgrx/src/nullable.rs

```diff
@@ -0,0 +1,523 @@
+
+    fn count_nulls(&self) -> usize;
+
+    /// Returns Some(true) if the element at `idx`` is valid (non-null),
+    /// or `None` if `idx` is out-of-bounds
+    /// Implementors should handle bounds-checking.
+    fn is_valid(&self, idx: Idx) -> Option<bool>;
+    /// Returns true if the element at idx is null,
+    /// or `None` if `idx` is out-of-bounds
+    /// Implementors should handle bounds-checking.
+    fn is_null(&self, idx: Idx) -> Option<bool> {
+        self.is_valid(idx).map(|v| !v)
+    }
+}
+
+pub trait SkippingNullLayout<Idx>: NullLayout<Idx>
+where
+    Idx: PartialEq + PartialOrd,
+{
+    /// For the given (linear) container index, returns the next index of the
+    /// underlying data buffer that should contain a valid value,
+    /// or `None` if we have reached the end of the container.
+    ///
+    /// Used to implement the skipping behavior expected with null-bitmap
+    /// -style arrays.
+    fn next_valid_idx(&self, idx: Idx) -> Option<Idx>;
+}
+
+/// All non-skipping null layouts. Marker trait for nullable containers that
+/// are nonetheless safe to iterate over linearly.
+pub trait ContiguousNullLayout<Idx>: NullLayout<Idx>
+where
+    Idx: PartialEq + PartialOrd,
+{
+}
+
+pub trait NullableContainer<'mcx, Idx, T>
+where
+    Idx: PartialEq + PartialOrd,
+{
+    type Layout: NullLayout<Idx>;
+
+    fn get_layout(&'mcx self) -> &'mcx Self::Layout;
+
+    /// For internal use - implement this over underlying data types
+    /// Used to implement NullableIter.
+    ///
+    /// Get the Valid value from the underlying data index of `idx`,
+    /// presumably after figuring out things like
+    fn get_raw(&'mcx self, idx: usize) -> T;
+}
+
+pub struct BitSliceNulls<'a>(pub &'a BitSlice<u8>);
+
+impl<'a> NullLayout<usize> for BitSliceNulls<'a> {
+    fn len(&self) -> usize {
+        BitSlice::<u8>::len(&self.0)
+    }
+
+    fn has_nulls(&self) -> bool {
+        self.0.not_all()
+    }
+
+    fn is_valid(&self, idx: usize) -> Option<bool> {
+        self.0.get(idx).map(|b| *b)
+    }
+    fn is_null(&self, idx: usize) -> Option<bool> {
+        self.0.get(idx).map(|b| !b)
+    }
+
+    fn count_nulls(&self) -> usize {
+        self.0.count_zeros()
+    }
+}
+
+impl<'a> SkippingNullLayout<usize> for BitSliceNulls<'a> {
+    fn next_valid_idx(&self, idx: usize) -> Option<usize> {
+        // Next elem (one after this) would be past the end
+        // of the container
+        if (idx + 1) >= self.0.len() {
+            return None;
+        }
+        let mut resulting_idx = 0;
+        for bit in &(*self.0)[(idx + 1)..] {
+            // Postgres nullbitmaps are 1 for "valid" and 0 for "null"
+            resulting_idx += (*bit) as usize;
+        }
+        Some(resulting_idx)
+    }
+}
+
+pub struct BoolSliceNulls<'a>(pub &'a [bool]);
+
+impl<'a> NullLayout<usize> for BoolSliceNulls<'a> {
+    fn len(&self) -> usize {
+        <[bool]>::len(&self.0)
+    }
+
+    fn has_nulls(&self) -> bool {
```

Comment:
Should doc-comment on cost of impl.

Reply:
Resolved in 1a3d5719b77c233fd96440597e3aa43d794847d4